### PR TITLE
Lift jagged -> padded dense forward / backward kernels from fbgemm_gpu

### PR DESCRIPTION
**Pull request by jbschlosser**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125946

PyTorch can't depend on `fbgemm_gpu` as a dependency because `fbgemm_gpu` already has a dependency on PyTorch. So this PR copy / pastes kernels from `fbgemm_gpu`:
* `dense_to_jagged_forward()` as CUDA registration for new ATen op `_padded_dense_to_jagged_forward()`
* `jagged_to_padded_dense_forward()` as CUDA registration for new ATen op `_jagged_to_padded_dense_forward()`

CPU impls for these new ATen ops will be added in a follow-up PR.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -14639,6 +14639,21 @@
     NestedTensorCUDA: NestedTensor_to_padded_tensor_cuda
   autogen: to_padded_tensor.out
 
+- func: _fbgemm_jagged_to_padded_dense_forward(Tensor values, Tensor[] offsets, SymInt[] max_lengths, float padding_value=0.0) -> Tensor
+  variants: function
+  dispatch:
+    CUDA: _fbgemm_jagged_to_padded_dense_forward
+
+- func: _fbgemm_jagged_to_padded_dense_backward(Tensor grad_output, Tensor[] offsets, SymInt total_L) -> Tensor
```
[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:14647]: There's a minor difference in whether `total_L` is required or not. I think in general, it's better for it not to be required because getting that value from `offsets[-1]` incurs a GPU sync.

So I think agree with you that we just need `jagged_to_padded_dense_forward` and `dense_to_jagged_forward`. I'll try removing the backward op.

[the remark this answers is not in the document]
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml]: Removed the backward op

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu [hunk 2]
```diff
@@ -1,3 +1,4 @@
+#include <cuda_fp16.h>
```
Comment (davidberard98) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:1]: I didn't review this file in much detail, I'm just assuming it's not significantly changed from fbgemm :)

Maybe can we add a code ptr to the fbgemm github repo so people can get more context?

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -14645,6 +14645,16 @@
...
 
+- func: _jagged_to_padded_dense_forward(Tensor values, Tensor[] offsets, SymInt[] max_lengths, float padding_value=0.0) -> Tensor
+  variants: function
+  dispatch:
+    CUDA: _fbgemm_jagged_to_padded_dense_forward
+
+- func: _padded_dense_to_jagged_forward(Tensor dense, Tensor[] offsets, SymInt? total_L=None) -> Tensor
+  variants: function
+  dispatch:
+    CUDA: _fbgemm_dense_to_jagged_forward_symint
```
Comment (davidberard98) [inline aten/src/ATen/native/native_functions.yaml:14656]: IIRC, some variants of `padded_dense_to_jagged_forward` also take a padding value, in case any of your sequences are longer than the sequence dimension of the dense tensor. I guess we probably don't care too much about this right now though?

**Review by davidberard98 — APPROVED** [review]
generally LGTM! I skimmed through the cuda implementations and they look fine from what I saw.

I guess we could also probably split this out from the stack to make sure there are no other failures masked by the failures introduced by the previous PR? (or we can wait if you prefer)

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu (hunk 2 above)
Comment (jbschlosser) [reply aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cu:1]: ah yes good call! will add an fbgemm pointer

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (jbschlosser) [reply aten/src/ATen/native/native_functions.yaml:14647]: this is a good point - I think for our initial op coverage purposes, we don't care too much. we may need to revisit this for more general usages though

**Comment by huydhn** [comment]
@clee2000 I have a PR here to exclude build jobs from flakiness check for now https://github.com/pytorch/test-infra/pull/5262 because the observation is that build job is usually stable enough

**Comment by jbschlosser** [comment]
@clee2000 / @huydhn what should I do to get signal on the failing windows build?

**Comment by jbschlosser** [comment]
for posterity: the windows build error appears to have been from use of a gcc-specific attribute:
```cpp
#define DEVICE_INLINE __device__ inline __attribute__((always_inline))
```

I changed this to use the cross-platform macro:
```cpp
#define DEVICE_INLINE __device__ C10_ALWAYS_INLINE
```

relanding.

**Comment by majing921201** [comment]
hi, @jbschlosser I read your PR for porting fbgemm_gpu kernels. That's great work. May I know is there a plan to port the autograd function for kernels. like https://github.com/pytorch/FBGEMM/blob/7d76ba77ba7e75930eb3ff16d76fe84224912318/fbgemm_gpu/src/jagged_tensor_ops/jagged_tensor_ops_autograd.cpp#L262
User can use `torch.dense_to_jagged` in model script for both inference and training
